### PR TITLE
gnrc_netif_ieee802154: rework `netif->state` to getter and flag

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

Low hanging fruit in the recv function. Removes direct access to the
netdev struct in favor of netdev API get calls. Reworking the access to the netdev struct in the send function requires a bit more work.

Impact on flash usage is 48 bytes on the samr21-xpro

### Issues/PRs references

Required for the changes proposed in #7736

### CODE REVIEW

File: sys/net/gnrc/netif/gnrc_netif_ieee802154.c [hunk 1]
```diff
@@ -95,7 +94,9 @@ static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
             gnrc_pktbuf_release(pkt);
             return NULL;
         }
-        if (!(state->flags & NETDEV_IEEE802154_RAW)) {
+        netopt_enable_t raw = NETOPT_DISABLE;
+        dev->driver->get(dev, NETOPT_RAWMODE, &raw, sizeof(raw));
+        if (!(raw == NETOPT_ENABLE)) {
```
Comment (gebart) [inline sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: Why not `!=`?

Comment (bergzand) [reply sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: Let me change that conditional instead of trying to answer your question :)

**Comment by bergzand** [comment]
> Does this add any measurable delay in processing packets?

AFAIK not significantly (ping delay didn't increase), but I have no way to accurately check this at the moment. I can probably build something quick and dirty tomorrow if you like.

File: sys/net/gnrc/netif/gnrc_netif_ieee802154.c (hunk 1 above)
Comment (bergzand) [reply sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: Any preference between `raw != NETOPT_ENABLE` or `raw == NETOPT_DISABLE`?

**Comment by bergzand** [comment]
@gebart Measuring with a samr21-xpro from the start to the end of the `recv()` function, comparing 2df6da25f5317556769209e6b21053f8dbbcb710 and the `HEAD` of this PR.
Using GPIO set/clear and a logic analyzer at 8Mhz, this measures in at a difference of 1.8us (386.6us vs 388.4us).

**Comment by gebart** [comment]
@bergzand nice testing, will you be able to do similar measurements in the layer PRs?

**Comment by bergzand** [comment]
> @bergzand nice testing, will you be able to do similar measurements in the layer PRs?

At the moment that's the idea yes. I'm rather curious about the impact of the changes versus the features it allows us to add.

File: sys/net/gnrc/netif/gnrc_netif_ieee802154.c [hunk 2]
```diff
@@ -95,7 +94,9 @@ static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
             gnrc_pktbuf_release(pkt);
             return NULL;
         }
-        if (!(state->flags & NETDEV_IEEE802154_RAW)) {
+        netopt_enable_t raw = NETOPT_DISABLE;
+        dev->driver->get(dev, NETOPT_RAWMODE, &raw, sizeof(raw));
+        if (raw == NETOPT_DISABLE) {
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: Mh... maybe have a flag for RAW mode in `gnrc_netif_t::flags` instead (of cause independent of IEEE 802.15.4) that is set on interface initialization and in `gnrc_netif_set_from_netdev()`. This way we may circumvent the (minimal) timing issue.

Comment (bergzand) [reply sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: Added the flag. I decided not to check if the flag is set on initialization of the netdev device as it really must not be enable by default in a netdev instance.

**Comment by bergzand** [comment]
@miri64 ping.

any remaining issues here? Is it okay if I squash this PR?

**Comment by bergzand** [comment]
Squashed and rebased

**Comment by bergzand** [comment]
> Why my measurement for non-`RAWMODE` is worse than [your original one](https://github.com/RIOT-OS/RIOT/pull/9564#issuecomment-404765331) I can't really explain (though I expected it to be better), and I'm not really sure anymore which state the PR originally was in (maybe I asked you to squash too early >.<). Let's merge for now. Maybe we can optimize timings later on by putting `proto` into `netif` altogether. IMHO it has no business being in the device driver anyway.

I should have documented my measurement setup a bit more, but I suspect that I've measured it with a nucleo-f446re and an mrf24j40, so different hardware causes a difference in measurement results. If you really like I can measure the diff again on a samr21-xpro and on the nucleo-mrf combo to see if I can confirm your results.

**Comment by miri64** [comment]
> I should have documented my measurement setup a bit more, but I suspect that I've measured it with a nucleo-f446re and an mrf24j40, so different hardware causes a difference in measurement results. If you really like I can measure the diff again on a samr21-xpro and on the nucleo-mrf combo to see if I can confirm your results.

If it doesn't cause you much trouble, I'd appreciate that.

**Pull request merged**